src/control/server/ctl_smd_rpc.go

```diff
@@ -44,74 +44,13 @@ func queryRank(reqRank uint32, engineRank ranklist.Rank) bool {
 
 func (svc *ControlService) querySmdDevices(ctx context.Context, req *ctlpb.SmdQueryReq, resp *ctlpb.SmdQueryResp) error {
 	for _, ei := range svc.harness.Instances() {
-		if !ei.IsReady() {
-			svc.log.Debugf("skipping not-ready instance %d", ei.Index())
-			continue
-		}
-
-		engineRank, err := ei.GetRank()
+		rResp, err := smdQueryEngine(ctx, ei, req)
 		if err != nil {
 			return err
 		}
-		if !queryRank(req.GetRank(), engineRank) {
-			continue
-		}
-
-		rResp := new(ctlpb.SmdQueryResp_RankResp)
-		rResp.Rank = engineRank.Uint32()
-
-		listDevsResp, err := ei.ListSmdDevices(ctx, new(ctlpb.SmdDevReq))
-		if err != nil {
-			return errors.Wrapf(err, "rank %d", engineRank)
-		}
-
-		if len(listDevsResp.Devices) == 0 {
-			rResp.Devices = nil
+		// If rank hasn't been requested, don't include its response.
+		if rResp != nil {
```

Comment:
Could rResp == nil even if err == nil?

Reply:
Yes, if ranks specified in the request don't match the engine then nil, nil will be returned. This is not behaviour that has been altered in this PR however.